### PR TITLE
NAS-124758 / 24.04 / NFS and directory service should wait for dependent services

### PR DESCRIPTION
**Pull request by mgrimesix**
Under some configurations, NFS can be dependent on nslcd and directory service, AD or LDAP (IPA) are nearly always dependent on the network subsystem.
Under some conditions and configurations, NFS and directory service would start before their dependent services.

NFS and nslcd are mostly managed by systemd.
To address NFS dependency on nslcd:  Added `After=nslcd` criteria to nfs-server unit file.

Directory Service is managed by middleware and, if enabled, is started as part of smb.configure.
To address Directory Service dependency on the network subsystem:  Added a sentinel file generated in the ix-netif unit file.  SMB `configure` will wait for the sentinel file before starting Directory Service.

Also did some cosmetic clean up of smb.py for flake8.

### CODE REVIEW

**Review by yocalebo — CHANGES_REQUESTED** [review]
Sentinel file won't work because of 2 reasons:
1. we have a `-t 120` (timeout) option in the `ExecStart` section of the ix-netif systemd unit file which means it's not guaranteed that network interfaces are up by the time it's done
2. I was just on a customer system where this systemd service "completed" but the NICs took another 30+ seconds to actually enter an `UP` state.

We'll probably have to have something in middleware that actually does something quite a bit more involved that actually checks to see if the interface that the service is bound to is actually `UP`. But even then, we'll have to have some timeout because we can't actually predict if the NIC will _EVER_ come `UP`. (Maybe switch port is down or someone pulled a cable)

**Comment by mgrimesix** [comment]
Unfortunate, but not surprising. 
A fail-safe timeout is already part of the 'wait'.   I'll craft a better network test.

**Comment by mgrimesix** [comment]
retest this please

**Comment by mgrimesix** [comment]
retest this please

File: src/middlewared/middlewared/plugins/smb.py [hunk 1]
```diff
@@ -494,6 +495,42 @@ async def ctdb_wait(self):
...
+    @private
+    async def netif_wait(self, timeout=120):
+        """
+        Wait for for the ix-netif sentinel file
+        and confirm connectivity with some targeted tests.
+        All must be completed before the timeout.
+        """
+        found_sentinel = False
+        while timeout >= 0 and not found_sentinel:
+            if os.path.exists(NETIF_COMPLETE_SENTINEL):
```
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/smb.py:507]:
```suggestion
            try:
                os.remove(NETIF_COMPLETE_SENTINEL)
                found_sentinel = True
            except FileNotFoundError:
                pass
```

File: src/middlewared/middlewared/plugins/smb.py [hunk 2]
```diff
@@ -494,6 +495,42 @@ async def ctdb_wait(self):
...
+                self.logger.warning('Failed to get netif completion sentinal.')
+            elif not found_sentinel:
+                await asyncio.sleep(1)
+
+        """
+        Confirm at least one network interface is UP
+        """
+        if_UP = False
+        while timeout >= 0 and not if_UP:
+            netif = await self.middleware.call('interface.query')
```
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/smb.py:522]: There is no reason to generate lists of interface names when we just need to detect if any of them are up. I suggest doing something like this
```suggestion
    while timeout >= 0:
        if any((
            i['state']['link_state'] == 'LINK_STATE_UP' for i in await self.middleware.call('interface.query')
        )):
            break
        else:
            timeout -= 1
            if timeout <= 0:
                self.logger.warning('Failed to detect any connected network interfaces.')
            else:
                await asyncio.sleep(1)
```

**Review by yocalebo — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/plugins/smb.py (hunk 1 above)
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/smb.py]: 👍

**Review by yocalebo — APPROVED** [review]
Thanks!

File: src/middlewared/middlewared/plugins/smb.py (hunk 2 above)
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/smb.py]: 👍

**Pull request merged**